dali/kernels/slice/slice_kernel_test.h

```diff
@@ -0,0 +1,191 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_SLICE_SLICE_KERNEL_TEST_H_
+#define DALI_KERNELS_SLICE_SLICE_KERNEL_TEST_H_
+
+#include <gtest/gtest.h>
+#include <vector>
+#include <string>
+#include "dali/kernels/test/tensor_test_utils.h"
+#include "dali/kernels/test/test_tensors.h"
+#include "dali/kernels/slice/slice_kernel_utils.h"
+
+namespace dali {
+namespace kernels {
+
+template <typename InputType_, typename OutputType_, std::size_t Dims_, std::size_t NumSamples_,
+          std::size_t DimSize_, typename SliceParamsGenerator_>
+struct SliceTestArgs {
+  using InputType = InputType_;
+  using OutputType = OutputType_;
+  static constexpr std::size_t Dims = Dims_;
+  static constexpr std::size_t NumSamples = NumSamples_;
+  static constexpr std::size_t DimSize = DimSize_;
+  using SliceArgsGenerator = SliceParamsGenerator_;
+};
+
+template <typename ExpectedShape>
+void AssertExpectedDimensions(const TensorShape<>& tensor_shape,
+                              const ExpectedShape& expected_shape) {
+  for (int j = 0; j < tensor_shape.size(); j++) {
+    ASSERT_EQ(expected_shape[j], tensor_shape[j]);
+  }
+}
+
+template <typename TestArgs>
+class SliceTest : public ::testing::Test {
+ public:
+  using InputType = typename TestArgs::InputType;
+  using OutputType = typename TestArgs::OutputType;
+  static constexpr std::size_t Dims = TestArgs::Dims;
+  static constexpr std::size_t NumSamples = TestArgs::NumSamples;
+  static constexpr std::size_t DimSize = TestArgs::DimSize;
+  using SliceArgsGenerator = typename TestArgs::SliceArgsGenerator;
+
+  void PrepareData(TestTensorList<InputType, Dims>& test_data) {
+    std::vector<int> sample_dims(Dims, DimSize);
+    TensorListShape<Dims> shape = uniform_list_shape<Dims>(NumSamples, sample_dims);
+    test_data.reshape(shape);
+
+    InputType num = 0;
+    auto seq_gen = [&num]() { return num++; };
+    Fill(test_data.cpu(), seq_gen);
+  }
+
+  void PrepareExpectedOutput(TestTensorList<InputType, Dims>& input_data,
+                             std::vector<SliceArgs<Dims>>& slice_args,
+                             TestTensorList<OutputType, Dims>& output_data) {
+    auto in = input_data.cpu();
+    std::vector<TensorShape<Dims>> output_shapes;
+    for (int i = 0; i < in.size(); i++) {
+      auto in_sample_shape = in.tensor_shape(i);
+      TensorShape<Dims> out_sample_shape(slice_args[i].shape);
+      auto& anchor = slice_args[i].anchor;
+
+      for (std::size_t d = 0; d < Dims; d++) {
+        ASSERT_TRUE(anchor[d] >= 0 && (anchor[d] + out_sample_shape[d]) <= in_sample_shape[d]);
```

Comment:
Nitpick: this could succeed with out-of-range anchors and negative output size - not that it's likely to happen.

Reply:
I think if we ever need to support this kind of semantics we can do it as a follow up